### PR TITLE
reducing poll latency with ioctl serial control

### PR DESCRIPTION
**Pull request by catch-twenty-two**
This pr fixes the latency on the serial port that is only noticeable when communicating at high speeds.  In turn this reduces the time it takes to write the mavlink ulog header on the companion computer side from 5-6 seconds to less than 1 second.  It may fix other latency issues that have been observed as well, but this hasn't been tested.

### CODE REVIEW

**Comment by jkflying** [comment]
Good catch! I guess the majority of this comes from setting `ASYNC_LOW_LATENCY`? The `O_NDELAY` seems like it might be dangerous on startup.

**Comment by jkflying** [comment]
This would also explain why we get the timesync issues when using mavlink-router, but not mavros: https://github.com/mavlink/mavros/blob/7c8b06ba17c9912628cd7c64f1f238e989863418/libmavconn/src/serial.cpp#L100

**Comment by catch-twenty-two** [comment]
> This would also explain why we get the timesync issues when using mavlink-router, but not mavros: https://github.com/mavlink/mavros/blob/7c8b06ba17c9912628cd7c64f1f238e989863418/libmavconn/src/serial.cpp#L100

Wow, look at that. Cool, that's promising then!

**Comment by catch-twenty-two** [comment]
> Good catch! I guess the majority of this comes from setting `ASYNC_LOW_LATENCY`? The `O_NDELAY` seems like it might be dangerous on startup.

I removed this, it actually does the same thing as the non-blocking setting in linux (they have the same value)

**Comment by RomanBapst** [comment]
@catch-twenty-two You the man! I will test this on my bench. Looking forward to finally see what the hell is going on during takeoff (4s of missing log :) )

File: src/mavlink-router/endpoint.cpp [hunk 1]
```diff
@@ -537,6 +541,16 @@ int UartEndpoint::open(const char *path)
         goto fail;
     }
 
+    /* For linux set poll latency at chip level */
+#ifdef __linux__
+    {
+        struct serial_struct serial_ctl;
+        ioctl(fd, TIOCGSERIAL, &serial_ctl);
```
Comment (bkueng) [inline src/mavlink-router/endpoint.cpp]: Can you check the return value?

**Comment by LorenzMeier** [comment]
@catch-twenty-two I'm designating you the official UART master at Auterion for now. Could you please make sure to cross-check if all of the settings that we do here are applied?
https://github.com/mavlink/c_uart_interface_example/blob/master/serial_port.cpp#L359-L496

A lot of developers have no idea what they are doing when it comes to UARTs, so every single time one gets used, we need to audit the config. I have seen this way too often in my career and further failure is not acceptable.

**Comment by LorenzMeier** [comment]
Also - I want to move maintenance to the mavlink org for mavlink-router.

**Comment by LorenzMeier** [comment]
https://github.com/intel/mavlink-router/issues/164

**Comment by jkflying** [comment]
I've tested this on the Up Core with FTDI < -- > PixRacer @ 921600, all of the `[ WARN] [1554972326.765113257]: TM : RTT too high for timesync: 22.37 ms.` messages are gone. 

Great work!

**Comment by catch-twenty-two** [comment]
> I've tested this on the Up Core with FTDI < -- > PixRacer @ 921600, all of the `[ WARN] [1554972326.765113257]: TM : RTT too high for timesync: 22.37 ms.` messages are gone.
> 
> Great work!

Nice, I'm glad it worked.

File: src/mavlink-router/endpoint.cpp (hunk 1 above)
Comment (catch-twenty-two) [reply src/mavlink-router/endpoint.cpp]: > @catch-twenty-two I'm designating you the official UART master at Auterion for now. Could you please make sure to cross-check if all of the settings that we do here are applied?
> https://github.com/mavlink/c_uart_interface_example/blob/master/serial_port.cpp#L359-L496
> 
> A lot of developers have no idea what they are doing when it comes to UARTs, so every single time one gets used, we need to audit the config. I have seen this way too often in my career and further failure is not acceptable.

Sound Good! Sure, I can take a look.

Comment (catch-twenty-two) [reply src/mavlink-router/endpoint.cpp]: Done

**Comment by catch-twenty-two** [comment]
> @catch-twenty-two I'm designating you the official UART master at Auterion for now. Could you please make sure to cross-check if all of the settings that we do here are applied?
> https://github.com/mavlink/c_uart_interface_example/blob/master/serial_port.cpp#L359-L496
> 
> A lot of developers have no idea what they are doing when it comes to UARTs, so every single time one gets used, we need to audit the config. I have seen this way too often in my career and further failure is not acceptable.

Sound Good! Sure, I can take a look.

File: src/mavlink-router/endpoint.cpp (hunk 1 above)
Comment (catch-twenty-two) [reply src/mavlink-router/endpoint.cpp]: upstream pr is here: https://github.com/intel/mavlink-router/pull/165

**Comment by jkflying** [comment]
@catch-twenty-two Did you test this on an ARM platform? Maybe PocketBeagle? We just did a test flight with this on Up Core and it worked out well, We got ~5 of the` [ WARN] : RTT too high for timesync` the whole flight, instead of 1 every 3 seconds. I'd like to get this merged, since it really helps the Avoidance side from a flight smoothness perspective.

**Comment by catch-twenty-two** [comment]
@jkflying I'll go ahead and test it on the arm today.

**Comment by catch-twenty-two** [comment]
@jkflying tested it with the pocket beagle. (which doesn't have a USB->Serial connection) works fine, also tested with my prolific serial adapter. Works Ok

**Pull request merged**